File: app/src/main/java/com/fpoly/project1/firebase/controller/ControllerCustomer.java
package com.fpoly.project1.firebase.controller;

import com.fpoly.project1.firebase.Firebase;
import com.fpoly.project1.firebase.model.Customer;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;

import java.util.ArrayList;
import java.util.Objects;

public class ControllerCustomer extends ControllerBase<Customer> {
    public ControllerCustomer() {
        super("table_customers");
    }

    /**
     * Broadcast object:
     * {
     * success: boolean, <-- response status, true if success, else false
     * value: any        <-- result, if any
     * error: String     <-- error message if failed
     * }
     * <p>
     * Firebase Realtime Database document
     * Please read: https://firebase.google.com/docs/database/android/read-and-write
     */

    @Override
    public boolean setSync(Customer value, boolean update) {
        DatabaseReference tableReference = Firebase.database.child(this.table);
        DatabaseReference rowReference;

        try {
            if (!update) {
                if (this.getAllSync().stream().anyMatch(account ->
                        account.emailAddress.equals(value.emailAddress) ||
                                account.gid.equals(value.gid) ||
                                account.fid.equals(value.fid))
                ) throw new Exception("Email đã tồn tại");

                rowReference = tableReference.push();

                // override ID
                value.__id = rowReference.getKey();

                Tasks.await(Firebase.database.child(this.table).child(Objects.requireNonNull(rowReference.getKey())).setValue(value));
            } else {
                rowReference = tableReference.child(value.__id);
                Tasks.await(rowReference.setValue(value));
            }

            return true;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    public String addSync(Customer value) {
        DatabaseReference tableReference = Firebase.database.child(this.table);
        DatabaseReference rowReference;

        try {
            if (this.getAllSync().stream().anyMatch(account ->
                    account.emailAddress.equals(value.emailAddress) ||
                            account.gid.equals(value.gid) ||
                            account.fid.equals(value.fid))
            ) throw new Exception("Email đã tồn tại");

            rowReference = tableReference.push();

            // override ID
            value.__id = rowReference.getKey();

            Tasks.await(Firebase.database.child(this.table).child(Objects.requireNonNull(rowReference.getKey())).setValue(value));

            return value.__id;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void setAsync(Customer value, boolean update, SuccessListener successListener, FailureListener failureListener) {

    }

    @Override
    public boolean removeSync(String referenceId) {
        try {
            Tasks.await(Firebase.database
                    .child(table)
                    .child(referenceId)
                    .setValue(null)
            );

            return true;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    @Override
    public void removeAsync(String referenceId, SuccessListener successListener, FailureListener failureListener) {

    }

    @Override
    public Customer getSync(String referenceId) {
        try {
            return Tasks.await(Firebase.database
                    .child(table)
                    .child(referenceId)
                    .get()
            ).getValue(Customer.class);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void getAsync(String referenceId, SuccessListener successListener, FailureListener failureListener) {

    }

    @Override
    public ArrayList<Customer> getAllSync() {
        try {
            ArrayList<Customer> list = new ArrayList<>();

            for (DataSnapshot dataSnapshot : Tasks.await(Firebase.database.child(table).get()).getChildren()) {
                list.add(dataSnapshot.getValue(Customer.class));
            }

            return list;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void getAllAsync(SuccessListener successListener, FailureListener failureListener) {

    }
}

File: app/src/main/java/com/fpoly/project1/firebase/controller/ControllerCustomer.kt
package com.fpoly.project1.firebase.controller

import android.util.Log
import com.fpoly.project1.firebase.Firebase
import com.fpoly.project1.firebase.model.Customer
import com.google.android.gms.tasks.Tasks
import com.google.firebase.FirebaseException
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseReference

class ControllerCustomer : ControllerBase<Customer>("table_customers") {
    /**
     * Broadcast object:
     * {
     * success: boolean, <-- response status, true if success, else false
     * value: any        <-- result, if any
     * error: String     <-- error message if failed
     * }
     *
     *
     * Firebase Realtime Database document
     * Please read: https://firebase.google.com/docs/database/android/read-and-write
     */
    override fun setSync(value: Customer, update: Boolean): Boolean {
        val tableReference = Firebase.database.child(table)
        val rowReference: DatabaseReference

        return try {
            if (!update) {
                val listUsers = this.getAllSync() ?: return false
                if (listUsers.stream()
                        .anyMatch { account: Customer? ->
                            account!!.emailAddress == value.emailAddress ||
                                    account.gid == value.gid ||
                                    account.fid == value.fid
                        }
                ) return false

                rowReference = tableReference.push()

                // override ID
                value.id = rowReference.key
                Tasks.await(
                    Firebase.database.child(table).child(value.id!!)
                        .setValue(value)
                )
            } else {
                rowReference = tableReference.child(value.id!!)
                Tasks.await(rowReference.setValue(value))
            }
            true
        } catch (e: FirebaseException) {
            Log.e(this.javaClass.simpleName, "Error while setting ${value.id}", e)

            false
        }
    }

    fun addSync(value: Customer): String? {
        val tableReference = Firebase.database.child(table)
        val rowReference: DatabaseReference
        return try {
            val listUsers = this.getAllSync() ?: return null
            if (listUsers.stream()
                    .anyMatch { account: Customer? ->
                        account!!.emailAddress == value.emailAddress ||
                                account.gid == value.gid ||
                                account.fid == value.fid
                    }
            ) {
                null
            } else {
                rowReference = tableReference.push()

                // override ID
                value.id = rowReference.key
                Tasks.await(
                    Firebase.database.child(table).child(value.id!!)
                        .setValue(value)
                )
                value.id
            }
        } catch (e: FirebaseException) {
            Log.e(this.javaClass.simpleName, "Error while adding ${value.id}", e)

            null
        }
    }

    fun addAsync(
        value: Customer,
        successListener: SuccessListener?,
        failureListener: FailureListener?
    ) {
        setAsync(value, false, successListener, failureListener)
    }

    override fun setAsync(
        value: Customer,
        update: Boolean,
        successListener: SuccessListener?,
        failureListener: FailureListener?
    ) {
        getAllAsync(
            successListener = object : SuccessListener() {
                override fun run(dataSnapshot: DataSnapshot?) {
                    val customers = ArrayList<Customer>()
                    if (dataSnapshot != null)
                        for (entry in dataSnapshot.children)
                            customers.add(entry.getValue(Customer::class.java)!!)

                    if (!update) {
                        if (customers.stream().anyMatch {
                                val matchEmail = it!!.emailAddress == value.emailAddress!!.trim()
                                val matchGid =
                                    (it.gid != null && value.gid != null && it.gid == value.gid)
                                val matchFid = (it.fid != null && value.fid != null && it.fid ==
                                        value.fid)

                                matchEmail || matchGid || matchFid
                            }) {
                            failureListener?.run(Exception("Entry already exists"))
                        } else {
                            val rowReference = Firebase.database.child(table).push()

                            // override ID
                            value.id = rowReference.key
                            Firebase.database.child(table).child(value.id!!)
                                .setValue(value)
                                .addOnCompleteListener {
                                    if (it.isSuccessful)
                                        successListener?.run(value.id)
                                    else
                                        failureListener?.run(it.exception)
                                }
                        }
                    } else {
                        Firebase.database.child(table).child(value.id!!).setValue(value)
                            .addOnCompleteListener {
                                if (it.isSuccessful)
                                    successListener?.run()
                                else
                                    failureListener?.run(it.exception)
                            }
                    }
                }
            },
            failureListener
        )
    }

    override fun removeSync(referenceId: String?): Boolean {
        return try {
            Tasks.await(
                Firebase.database
                    .child(table)
                    .child(referenceId!!)
                    .setValue(null)
            )
            true
        } catch (e: FirebaseException) {
            Log.e(this.javaClass.simpleName, "Error while removing $referenceId", e)

            false
        }
    }

    override fun getSync(referenceId: String?): Customer? {
        return try {
            Tasks.await(
                Firebase.database
                    .child(table)
                    .child(referenceId!!)
                    .get()
            ).getValue(Customer::class.java)
        } catch (e: FirebaseException) {
            Log.e(this.javaClass.simpleName, "Error while getting $referenceId", e)

            null
        }
    }

    override fun getAllSync(): ArrayList<Customer>? {
        return try {
            val list = ArrayList<Customer>()
            for (dataSnapshot in Tasks.await(Firebase.database.child(table).get()).children) {
                dataSnapshot.getValue(Customer::class.java)?.let { list.add(it) }
            }

            list
        } catch (e: FirebaseException) {
            Log.e(this.javaClass.simpleName, "Error while getting all entries", e)

            null
        }
    }
}
